# Prepare the weather data

In [0]:
import datetime
from pyspark.sql.types import *
from pyspark.sql.functions import unix_timestamp
import math
from pyspark.sql import functions as F

To begin, take a look at the `flight_weather_with_airport_code` data that was imported to get a sense of the data we will be working with.

In [0]:
%sql
select * from flight_weather_with_airport_code

Year Month Day Time TimeZone SkyCondition Visibility WeatherType DryBulbFarenheit DryBulbCelsius WetBulbFarenheit WetBulbCelsius DewPointFarenheit DewPointCelsius RelativeHumidity WindSpeed WindDirection ValueForWindCharacter StationPressure PressureTendency PressureChange SeaLevelPressure RecordType HourlyPrecip Altimeter AirportCode DISPLAY_AIRPORT_NAME LATITUDE LONGITUDE 2013 4 1 56 -4 FEW018 SCT044 BKN070 10.00 -RA 76 24.4 74 23.3 73 22.8 90 13 080 null 30.06 null null 30.06 AA T 30.07 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 156 -4 FEW037 SCT070 10.00 null 76 24.4 73 22.5 71 21.7 85 10 090 null 30.05 6 17 30.05 AA null 30.06 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 256 -4 FEW037 SCT070 10.00 null 76 24.4 73 22.5 71 21.7 85 9 100 null 30.03 null null 30.03 AA null 30.04 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 356 -4 FEW025 SCT070 10.00 null 76 24.4 72 22.2 70 21.1 82 9 100 null 30.02 null null 30.03 AA null 30.03 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 456 -4 FEW025 10.00 null 76 24.4 72 22.2 70 21.1 82 7 110 null 30.03 5 4 30.04 AA null 30.04 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 556 -4 FEW025 SCT080 10.00 null 76 24.4 71 21.8 69 20.6 79 7 100 null 30.04 null null 30.05 AA null 30.05 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 656 -4 FEW028 BKN080 10.00 null 77 25.0 71 21.7 68 20.0 74 9 110 null 30.07 null null 30.07 AA null 30.08 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 756 -4 FEW028 BKN080 10.00 null 79 26.1 72 22.4 69 20.6 72 13 100 null 30.09 3 20 30.10 AA null 30.10 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 856 -4 FEW030 BKN080 10.00 null 82 27.8 73 22.9 69 20.6 65 14 100 21 30.11 null null 30.11 AA null 30.12 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 956 -4 SCT035 BKN090 10.00 null 83 28.3 74 23.0 69 20.6 63 16 090 23 30.11 null null 30.12 AA null 30.12 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1056 -4 SCT035 BKN090 10.00 null 84 28.9 74 23.5 70 21.1 63 17 080 24 30.12 1 8 30.12 AA null 30.13 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1156 -4 FEW026 BKN090 10.00 null 84 28.9 74 23.5 70 21.1 63 16 080 25 30.09 null null 30.10 AA null 30.10 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1256 -4 FEW028 BKN090 10.00 null 86 30.0 75 23.9 70 21.1 59 16 080 25 30.07 null null 30.08 AA null 30.08 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1356 -4 FEW040 BKN090 10.00 null 86 30.0 76 24.2 71 21.7 61 20 080 25 30.05 8 23 30.05 AA null 30.06 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1456 -4 FEW033 SCT042 SCT120 10.00 null 86 30.0 75 23.9 70 21.1 59 18 070 25 30.03 null null 30.03 AA null 30.04 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1556 -4 FEW045 SCT065 SCT120 10.00 null 85 29.4 74 23.0 68 20.0 57 20 090 26 30.02 null null 30.02 AA null 30.03 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1656 -4 FEW033 SCT048 SCT055 10.00 null 83 28.3 73 22.7 68 20.0 61 15 090 28 30.02 5 8 30.03 AA null 30.03 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1756 -4 FEW039 SCT085 SCT120 10.00 null 81 27.2 73 22.7 69 20.6 67 18 080 23 30.03 null null 30.03 AA null 30.04 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1856 -4 FEW025 SCT049 SCT070 10.00 null 80 26.7 73 22.5 69 20.6 69 11 110 21 30.05 null null 30.05 AA null 30.06 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 1956 -4 FEW049 SCT060 10.00 null 78 25.6 71 21.8 68 20.0 71 11 090 null 30.06 3 11 30.06 AA null 30.07 SJU Luis Munoz Marin International 18.43944444 -66.00222222 2013 4 1 2056 -4 FEW049 SCT070 10.00 null 78 25.6 73 22.5 70 21.1 77 9 090 null 30.07 null null 30.07 AA null 30.08 SJU Luis Mun

Next, count the number of records so we know how many rows we are working with.

In [0]:
%sql
select count(*) from flight_weather_with_airport_code

count(1) 406516

Observe that this data set has 406,516 rows and 29 columns. For this model, we are going to focus on predicting delays using WindSpeed (in MPH), SeaLevelPressure (in inches of Hg), and HourlyPrecip (in inches). We will focus on preparing the data for those features.

Let's start out by taking a look at the **WindSpeed** column. You may scroll through the values in the table above, but reviewing just the distinct values will be faster.

In [0]:
%sql
select distinct WindSpeed from flight_weather_with_airport_code

WindSpeed 7 51 15 11 29 3 30 34 8 22 28 16 0 47 null 43 5 31 18 17 26 46 M 6 19 23 41 40 38 25 44 33 9 24 32 20 36 10 37 39 13 14 21 45 27 53 62

Try clicking on the **WindSpeed** column header to sort the list by ascending and then by descending order. Observe that the values are all numbers, with the exception of some having `null` values and a string value of `M` for Missing. We will need to ensure that we remove any missing values and convert WindSpeed to its proper type as a numeric feature.

Next, let's take a look at the **SeaLevelPressure** column in the same way, by listing its distinct values.

In [0]:
%sql
select distinct SeaLevelPressure from flight_weather_with_airport_code

SeaLevelPressure 29.68 29.45 30.43 29.58 30.59 30.13 30.66 29.39 30.17 29.61 30.41 29.42 30.42 30.34 30.52 30.71 29.20 30.25 30.09 29.10 29.83 30.54 29.54 29.63 30.45 30.24 29.40 30.33 29.73 29.34 29.13 29.24 30.18 29.99 29.57 30.00 29.69 29.09 30.46 30.68 29.37 29.71 29.49 29.60 29.93 30.02 30.57 29.85 30.22 29.89 29.64 29.30 29.36 29.31 30.27 30.56 30.50 29.28 29.82 29.27 M 29.70 29.11 30.28 29.97 29.47 29.38 30.04 29.88 30.49 29.77 30.35 29.84 29.44 30.31 29.95 29.86 30.38 29.50 30.76 29.26 30.74 30.14 30.53 30.58 30.07 30.60 30.63 30.23 30.08 29.74 30.62 29.12 29.79 30.37 29.59 30.64 29.53 30.03 30.72 29.32 30.29 30.11 30.21 30.30 30.39 30.05 30.06 29.17 29.87 29.56 29.65 29.19 30.36 30.19 30.44 30.67 29.80 30.32 29.81 29.72 30.69 30.73 29.94 30.75 29.78 29.51 29.75 29.66 29.29 29.48 30.65 29.14 30.01 30.55 29.55 29.76 30.10 30.15 30.70 29.46 29.91 29.98 29.43 30.40 30.51 29.67 29.52 29.33 30.12 29.41 29.35 29.96 30.16 29.90 30.48 29.22 29.92 30.61 29.62 30.47 30.20 30.26 29.23 29.16 29.05 29.21 29.08 29.25 29.15 29.06 29.18 29.07 29.03 29.01 28.96 29.00 28.99 28.95 29.02 29.04 28.97 28.98

Like you did before, click on the **SeaLevelPressure** column header to sort the values in ascending and then descending order. Observe that many of the features are of a numeric value (e.g., 29.96, 30.01, etc.), but some contain the string value of M for Missing. We will need to replace this value of "M" with a suitable numeric value so that we can convert this feature to be a numeric feature.

Finally, let's observe the **HourlyPrecip** feature by selecting its distinct values.

In [0]:
%sql
select distinct HourlyPrecip from flight_weather_with_airport_code

HourlyPrecip 0.55 0.07 0.75 0.59 1.53 0.32 0.03 0.11 1.23 0.60 0.31 0.73 0.14 0.74 0.78 0.57 0.25 0.99 0.36 0.30 1.50 1.11 0.68 0.82 1.28 0.06 0.95 1.02 0.72 0.90 null 1.27 1.63 2.00 T 0.76 2.40 0.21 0.15 0.80 1.48 0.87 0.39 0.35 0.42 0.08 0.71 0.51 0.79 0.02 0.86 0.94 0.12 0.45 1.14 0.84 0.89 0.53 1.78 0.22 0.33 1.64 1.03 1.36 0.10 1.37 1.18 1.65 0.77 1.09 0.38 0.49 0.52 0.16 1.12 0.05 0.40 1.25 0.62 0.65 2.29 0.37 0.50 1.89 0.91 0.24 0.70 0.26 1.55 0.19 0.09 1.08 0.13 0.34 3.03 0.48 0.43 0.66 2.04 0.83 0.27 0.56 0.04 0.18 0.58 2.45 0.97 1.99 0.93 0.29 0.20 0.41 0.85 2.17 2.36 1.90 0.67 0.28 1.77 0.01 1.26 0.61 0.23 1.34 0.98 0.92 1.20 0.64 1.42 1.22 1.07 1.06 1.80 1.00 0.17 0.63 1.10 0.47 0.46 0.44 1.05 0.81 0.54 0.88 0.69 2.49 1.96 1.68 1.41 1.85 1.01 1.73 2.89 2.86 0.96 2.21 1.75 1.88 1.47 1.67 1.40 1.17 1.24 1.13 1.62 1.84 1.21 2.11 1.83 1.39 1.72 2.70 4.00 1.29 1.04 1.79 1.16 1.30 1.38 1.54 1.91 1.57

Click on the column header to sort the list and ascending and then descending order. Observe that this column contains mostly numeric values, but also `null` values and values with `T` (for Trace amount of rain). We need to replace T with a suitable numeric value and convert this to a numeric feature.

## Clean up weather data

To preform our data cleanup, we will execute a Python script, in which we will perform the following tasks:

* WindSpeed: Replace missing values with 0.0, and “M” values with 0.005
* HourlyPrecip: Replace missing values with 0.0, and “T” values with 0.005
* SeaLevelPressure: Replace “M” values with 29.92 (the average pressure)
* Convert WindSpeed, HourlyPrecip, and SeaLevelPressure to numeric columns
* Round “Time” column down to the nearest hour, and add value to a new column named “Hour”
* Eliminate unneeded columns from the dataset

Let's begin by creating a new DataFrame from the table. While we're at it, we'll pare down the number of columns to just the ones we need (AirportCode, Month, Day, Time, WindSpeed, SeaLevelPressure, HourlyPrecip).

In [0]:
dfWeather = spark.sql("select AirportCode, cast(Month as int) Month, cast(Day as int) Day, cast(Time as int) Time, WindSpeed, SeaLevelPressure, HourlyPrecip from flight_weather_with_airport_code")

In [0]:
dfWeather.show()

+-----------+-----+---+----+---------+----------------+------------+
|AirportCode|Month|Day|Time|WindSpeed|SeaLevelPressure|HourlyPrecip|
+-----------+-----+---+----+---------+----------------+------------+
|        SJU|    4|  1|  56|       13|           30.06|           T|
|        SJU|    4|  1| 156|       10|           30.05|        null|
|        SJU|    4|  1| 256|        9|           30.03|        null|
|        SJU|    4|  1| 356|        9|           30.03|        null|
|        SJU|    4|  1| 456|        7|           30.04|        null|
|        SJU|    4|  1| 556|        7|           30.05|        null|
|        SJU|    4|  1| 656|        9|           30.07|        null|
|        SJU|    4|  1| 756|       13|           30.10|        null|
|        SJU|    4|  1| 856|       14|           30.11|        null|
|        SJU|    4|  1| 956|       16|           30.12|        null|
|        SJU|    4|  1|1056|       17|           30.12|        null|
|        SJU|    4|  1|1156|      

Review the schema of the dfWeather DataFrame

In [0]:
print(dfWeather.dtypes)

[('AirportCode', 'string'), ('Month', 'int'), ('Day', 'int'), ('Time', 'int'), ('WindSpeed', 'string'), ('SeaLevelPressure', 'string'), ('HourlyPrecip', 'string')]


In [0]:

# Round Time down to the next hour, since that is the hour for which we want to use flight data. Then, add the rounded Time to a new column named "Hour", and append that column to the dfWeather DataFrame.
df = dfWeather.withColumn('Hour', F.floor(dfWeather['Time']/100))

# Replace any missing HourlyPrecip and WindSpeed values with 0.0
df = df.fillna('0.0', subset=['HourlyPrecip', 'WindSpeed'])

# Replace any WindSpeed values of "M" with 0.005
df = df.replace('M', '0.005', 'WindSpeed')

# Replace any SeaLevelPressure values of "M" with 29.92 (the average pressure)
df = df.replace('M', '29.92', 'SeaLevelPressure')

# Replace any HourlyPrecip values of "T" (trace) with 0.005
df = df.replace('T', '0.005', 'HourlyPrecip')

# Be sure to convert WindSpeed, SeaLevelPressure, and HourlyPrecip columns to float
# Define a new DataFrame that includes just the columns being used by the model, including the new Hour feature
dfWeather_Clean = df.select('AirportCode', 'Month', 'Day', 'Hour', df['WindSpeed'].cast('float'), df['SeaLevelPressure'].cast('float'), df['HourlyPrecip'].cast('float'))


Now let's take a look at the new `dfWeather_Clean` DataFrame.

In [0]:
display(dfWeather_Clean)

AirportCode Month Day Hour WindSpeed SeaLevelPressure HourlyPrecip SJU 4 1 0 13.0 30.06 0.005 SJU 4 1 1 10.0 30.05 0.0 SJU 4 1 2 9.0 30.03 0.0 SJU 4 1 3 9.0 30.03 0.0 SJU 4 1 4 7.0 30.04 0.0 SJU 4 1 5 7.0 30.05 0.0 SJU 4 1 6 9.0 30.07 0.0 SJU 4 1 7 13.0 30.1 0.0 SJU 4 1 8 14.0 30.11 0.0 SJU 4 1 9 16.0 30.12 0.0 SJU 4 1 10 17.0 30.12 0.0 SJU 4 1 11 16.0 30.1 0.0 SJU 4 1 12 16.0 30.08 0.0 SJU 4 1 13 20.0 30.05 0.0 SJU 4 1 14 18.0 30.03 0.0 SJU 4 1 15 20.0 30.02 0.0 SJU 4 1 16 15.0 30.03 0.0 SJU 4 1 17 18.0 30.03 0.0 SJU 4 1 18 11.0 30.05 0.0 SJU 4 1 19 11.0 30.06 0.0 SJU 4 1 20 9.0 30.07 0.0 SJU 4 1 21 10.0 30.09 0.01 SJU 4 1 22 9.0 30.08 0.005 SJU 4 1 23 7.0 30.07 0.0 SJU 4 2 0 9.0 30.05 0.0 SJU 4 2 1 9.0 30.03 0.0 SJU 4 2 2 5.0 30.03 0.0 SJU 4 2 3 6.0 30.02 0.0 SJU 4 2 4 3.0 30.02 0.0 SJU 4 2 5 8.0 30.03 0.0 SJU 4 2 6 0.0 30.04 0.0 SJU 4 2 7 10.0 30.06 0.0 SJU 4 2 8 8.0 30.06 0.0 SJU 4 2 9 9.0 30.06 0.0 SJU 4 2 10 11.0 30.05 0.0 SJU 4 2 11 13.0 30.03 0.0 SJU 4 2 12 13.0 30.01 0.0 SJU 4 2 13 13.0 29.99 0.0 SJU 4 2 14 10.0 29.98 0.0 SJU 4 2 15 14.0 29.98 0.0 SJU 4 2 16 8.0 29.99 0.0 SJU 4 2 17 9.0 29.99 0.0 SJU 4 2 18 8.0 30.01 0.0 SJU 4 2 19 5.0 30.03 0.0 SJU 4 2 20 0.0 30.04 0.01 SJU 4 2 21 3.0 30.06 0.0 SJU 4 2 22 3.0 30.06 0.0 SJU 4 2 23 0.0 30.06 0.0 SJU 4 3 0 0.0 30.04 0.0 SJU 4 3 1 0.0 30.01 0.0 SJU 4 3 2 3.0 30.01 0.0 SJU 4 3 3 0.0 30.01 0.0 SJU 4 3 4 0.0 30.01 0.0 SJU 4 3 5 3.0 30.03 0.0 SJU 4 3 6 0.0 30.04 0.0 SJU 4 3 7 0.0 30.06 0.0 SJU 4 3 8 0.0 30.08 0.0 SJU 4 3 9 8.0 30.07 0.0 SJU 4 3 10 0.0 30.06 0.0 SJU 4 3 11 11.0 30.05 0.0 SJU 4 3 12 15.0 30.02 0.0 SJU 4 3 13 20.0 30.0 0.0 SJU 4 3 14 16.0 29.98 0.0 SJU 4 3 15 16.0 29.97 0.0 SJU 4 3 16 20.0 29.97 0.0 SJU 4 3 17 10.0 29.98 0.0 SJU 4 3 18 7.0 30.0 0.0 SJU 4 3 19 11.0 30.02 0.0 SJU 4 3 20 8.0 30.03 0.0 SJU 4 3 21 5.0 30.05 0.0 SJU 4 3 22 5.0 30.06 0.0 SJU 4 3 23 3.0 30.05 0.0 SJU 4 4 0 0.0 30.03 0.0 SJU 4 4 1 0.0 30.03 0.0 SJU 4 4 2 0.0 30.02 0.0 SJU 4 4 3 0.0 30.01 0.0 SJU 4 4 4 0.0 30.02 0.0 SJU 4 4 5 0.0 30.04 0.0 SJU 4 4 6 0.0 30.06 0.0 SJU 4 4 7 7.0 30.08 0.0 SJU 4 4 8 10.0 30.09 0.0 SJU 4 4 9 14.0 30.09 0.0 SJU 4 4 10 16.0 30.09 0.0 SJU 4 4 11 21.0 30.08 0.0 SJU 4 4 12 21.0 30.06 0.0 SJU 4 4 13 20.0 30.03 0.0 SJU 4 4 14 22.0 30.02 0.0 SJU 4 4 15 17.0 30.01 0.0 SJU 4 4 16 16.0 30.01 0.0 SJU 4 4 17 20.0 30.01 0.0 SJU 4 4 18 16.0 30.02 0.0 SJU 4 4 19 14.0 30.03 0.0 SJU 4 4 20 9.0 30.06 0.0 SJU 4 4 21 10.0 30.07 0.0 SJU 4 4 22 9.0 30.07 0.0 SJU 4 4 23 5.0 30.06 0.01 SJU 4 5 0 5.0 30.05 0.01 SJU 4 5 1 6.0 30.02 0.02 SJU 4 5 2 7.0 30.0 0.02 SJU 4 5 3 7.0 29.99 0.03 SJU 4 5 4 5.0 29.99 0.0 SJU 4 5 5 5.0 30.02 0.02 SJU 4 5 6 11.0 30.03 0.005 SJU 4 5 7 10.0 30.05 0.0 SJU 4 5 8 8.0 30.07 0.005 SJU 4 5 9 7.0 30.09 0.005 SJU 4 5 10 8.0 30.08 0.005 SJU 4 5 11 10.0 30.06 0.0 SJU 4 5 12 13.0 30.03 0.0 SJU 4 5 13 18.0 30.0 0.0 SJU 4 5 14 18.0 29.97 0.0 SJU 4 5 15 17.0 29.96 0.0 SJU 4 5 16 15.0 29.96 0.01 SJU 4 5 17 16.0 29.97 0.01 SJU 4 5 18 10.0 29.98 0.0 SJU 4 5 19 7.0 30.01 0.0 SJU 4 5 20 5.0 30.02 0.01 SJU 4 5 21 3.0 30.03 0.0 SJU 4 5 22 0.0 30.03 0.02 SJU 4 5 23 0.0 30.02 0.0 SJU 4 6 0 0.0 30.01 0.0 SJU 4 6 1 0.0 29.99 0.0 SJU 4 6 2 0.0 29.98 0.0 SJU 4 6 3 0.0 29.98 0.0 SJU 4 6 4 0.0 29.98 0.0 SJU 4 6 5 0.0 30.0 0.0 SJU 4 6 6 0.0 30.03 0.01 SJU 4 6 7 0.0 30.04 0.0 SJU 4 6 8 0.0 30.04 0.0 SJU 4 6 9 6.0 30.04 0.0 SJU 4 6 10 10.0 30.04 0.0 SJU 4 6 11 8.0 30.03 0.0 SJU 4 6 12 16.0 30.01 0.0 SJU 4 6 13 17.0 29.97 0.0 SJU 4 6 14 9.0 29.96 0.0 SJU 4 6 15 9.0 29.95 0.0 SJU 4 6 16 8.0 29.95 0.0 SJU 4 6 17 6.0 29.97 0.0 SJU 4 6 18 5.0 29.99 0.02 SJU 4 6 19 5.0 30.01 0.0 SJU 4 6 20 5.0 30.03 0.0 SJU 4 6 21 3.0 30.04 0.0 SJU 4 6 22 5.0 30.05 0.0 SJU 4 6 23 6.0 30.03 0.0 SJU 4 7 0 5.0 30.02 0.0 SJU 4 7 1 3.0 30.0 0.02 SJU 4 7 2 3.0 29.98 0.01 SJU 4 7 3 0.0 29.97 0.0 SJU 4 7 4 0.0 29.98 0.0 SJU 4 7 5 0.0 30.0 0.01 SJU 4 7 6 3.0 30.01 0.04 SJU 4 7 7 3.0 30.03 0.005 SJU 4 7 8 6.0 30.05 0.0 SJU 4 7 9 8.0 30.06 0.0 SJU 4 7 10 11.0 29.92 0.0 SJU 4 7 10 11.0 29.92 0.0 SJU 4 7 10 13.0 30.05

Observe that the new DataFrame only has 7 columns. Also, the WindSpeed, SeaLevelPressure, and HourlyPrecip fields are all numeric and contain no missing values. To ensure they are indeed numeric, we can take a look at the DataFrame's schema.

In [0]:
print(dfWeather_Clean.dtypes)

[('AirportCode', 'string'), ('Month', 'int'), ('Day', 'int'), ('Hour', 'bigint'), ('WindSpeed', 'float'), ('SeaLevelPressure', 'float'), ('HourlyPrecip', 'float')]


Now let's persist the cleaned weather data to a persistent global table.

In [0]:
dfWeather_Clean.write.mode("overwrite").save("/mnt/sparkcontainer/Silver/flight_weather_clean")

In [0]:
%sql
DROP TABLE IF EXISTS flight_weather_clean;

CREATE TABLE flight_weather_clean
USING DELTA LOCATION '/mnt/sparkcontainer/Silver/flight_weather_clean'

In [0]:
dfWeather_Clean.select("*").count()

Out[15]: 406516